cleverhans/model.py

```diff
@@ -0,0 +1,221 @@
+from abc import ABCMeta
+
+
+class Model(object):
+    """
+    An abstract interface for model wrappers that exposes model symbols
+    needed for making an attack. This abstraction removes the dependency on
+    any specific neural network package (e.g. Keras) from the core
+    code of CleverHans. It can also simplify exposing the hidden features of a
+    model when a specific package does not directly expose them. For an
+    example of why this is useful, see the paper "Adversarial Manipulation
+    of Deep Representations" (https://arxiv.org/abs/1511.05122).
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
```

Comment:
It is true that this is a weird interface but that's only because as far as I know python does not have a nice way of casting an object to a subclass. @npapernot's commit 45bcf6dd3c8de22e747a6efefd6ead2ddbf6a7d3, address this issue by adding a separate `set_model` method. That way we can have models that implement `Model` without inheriting from any other class, e.g. Ian's pure TF model. But I think it still makes sense to have `__init__(self, model)` for `KerasModelWrapper` in addition to `__init(self)`. It can simply make a call to `set_model`.